Threading

rcombs opened the pull request:
Replaces #636, as it'd been through several major cache redesigns and the thread was too long to be practical review-wise.

rcombs committed:
build: detect threads and atomics

rcombs committed:
ass_font(select): add thread-safety

rcombs committed:
ass_render: add support for per-event threaded rendering

This adds ass_set_threads(), which sets the number of threads to use.
By default, libass will select a thread count automatically based on
the available number of CPU cores on the system.
Calling ass_set_threads() with any value will also opt out of internal
locking around log calls, which improves performance.
To opt out of log locking while allowing libass to autoselect a thread count,
pass 0 as the thread count.
To opt out of threading altogether, pass 1 as the thread count.
If more than 1 thread is in use, log callbacks may be made from threads
spawned internally by libass, rather than the one calling ass_render_frame().

rcombs committed:
ass_font: add locking routines

rcombs committed:
ass_font: access FT_Face thread-safely

TheOneric committed:
ci/gha: add a TSAN build

rcombs committed:
build: require fribidi 0.19.7

Versions prior to this were not thread-safe by default.

rcombs committed:
ass_threading: add pthread/atomics support header

Also enables warnings on implicit atomic ops (they're slower, and aren't
supported by the compat shims).

rcombs committed:
build: remove explicit -std=gnu99; prefer C11 if available

This is automatically set by AM_PROG_CC if needed,
and potentially to a newer version (e.g. C11, which we need for atomics).

rcombs committed:
ass_utils: add locking around log callback

rcombs committed:
ass_shaper: access FT_Face thread-safely

rcombs committed:
ass_cache: introduce CacheClient API

This will later be used to shard promotion and locking per-thread

rcombs committed:
ass_cache: split key and value destructors

rcombs committed:
ass_cache: introduce lockfree thread-safe architecture

rcombs committed:
tools: opt into thread-safe logging

TheOneric commented on Makefile.am:
```diff
@@ -1,6 +1,6 @@
 ACLOCAL_AMFLAGS = -I m4
 AM_CPPFLAGS = -I$(top_srcdir)/libass
-AM_CFLAGS = -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter \
+AM_CFLAGS = -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter \
```
No longer requesting an explicit standard version has undesireable effects:
 - our expected baseline is no longer documented at all
 - we now depend on compiler and/or autotools defaults which might change at any time and differ across compilers
 - with the current C11 default we’ll no longer test our actual default in CI, nor on developer machines
 
 This makes it likely we’ll accidentally break our supposed baseline at some point by inadvertenly using a newer feature *(see e.g. what happened with an `{}` between 0.17.1 and 0.17.2 because compilers didn’t warn about it)*.
 
 On the other hand, the reason for dropping this is some libcs otherwise hiding atomics; which libcs do that?
 
I think instead of just dropping the standard version we should do one of the following *(ordered by my personal preference)*:
- continue to explicitly request GNU99 **if** no libcs actually hide atomics
- raise our baseline to C11 and explicitly request GNU/ISO C11
- explicitly request _either_ GNU/ISO C11 _or_ GNU/ISO C99 and make sure at least one CI job always uses C99
- drop the explicit version like now, but ensure one CI job explicitly runs with our baseline version
 
*(Having a meson buildsystem and meson not _yet_ being changed actually helps out here, since it still explicitly requests ISO C99, but once meson will change too so this is only temporary)*

rcombs replied:
`AC_PROG_CC` automatically selects the latest known non-strict-conformance (i.e. `gnu`) standard version the compiler supports. If we want, we can check `ac_prog_cc_stdc` and warn if it's `c89`, but I don't really see the need; if an older compiler fails on our code due to not supporting the minimum required standard, the user will be notified by, uh… compilation failing (which is already what happens now, just slightly faster via `-std=gnu99` being unrecognized). I think the only actual potential issue here is that, say, a future C version could come along and remove something we use? But generally the C standard provides years of deprecation notice, so I'm not really worried about that either.

Regardless of whether atomics are hidden in C99 mode, I see no benefit in unconditionally explicitly requesting an older version than what we're using features from.

Running a build with an explicit `-std=gnu[earliest supported version]` seems reasonable, though I think it'd be more useful to do one with the earliest *gcc version* we're targeting (which is a much more meaningful test than using a newer compiler in an older mode).

Documenting the minimum requirements somewhere seems reasonable (README?); as of this PR it's "99, or 11 for threading", and I expect there's a good chance it'll move to "11, unless we have an explicit compat shim for your system" in the near future (assuming we go forwards with requiring atomics for publicly-refcounted frames).

TheOneric commented on libass/ass_utils.c:
```diff
@@ -116,10 +117,20 @@ void *ass_try_realloc_array(void *ptr, size_t nmemb, size_t size)
+#if ENABLE_THREADS
+    if (!priv->thread_safe_cb)
+        pthread_mutex_lock(&priv->log_mutex);
+#endif
```
Potential cosmetic issue: our default log handler uses several printf calls and is listed as thread safe. Individual calls from different messages might end up interleaved.

Strictly speaking, albeit only a theoretical concern for any remotely relevant platform as far as i’m aware: ISO C99 to my knowledge doesn't guarantee thread-safety of `(f))printf`. [POSIX explicitly guarantees this](https://pubs.opengroup.org/onlinepubs/9699919799/functions/flockfile.html) and provides `_unlocked` non-thread-safe variants for when custom locking is used and print perf relevant.

rcombs replied:
Yeah, this interleaving does happen in practice, but I'm not particularly concerned about it (since as you say it's cosmetic). I've found the overhead of locking to actually be meaningful in some cases (esp. when people put nonstandard stuff in the `Effect` field, since that logs on every frame).

TheOneric replied:
yeah, i’m not too concerned either; mostly just brought it up again since the old thread ended without a conclusion on `flockfile`.  
If interleaving turns out too annoying in practice we can still change this later

TheOneric reviewed (commented):
Moving over still relevant comments from the old PR

TheOneric commented on profile/profile.c:
```diff
@@ -57,6 +57,7 @@ static void init(int frame_w, int frame_h)
         exit(1);
     }
 
+    ass_set_threads(ass_renderer, 0);
```
At least `fuzz`, possibly all other utilities too, should opt into thread-safe logging

rcombs replied:
Done.

TheOneric commented on configure.ac:
```diff
@@ -80,6 +84,40 @@ AC_SEARCH_LIBS([lrint], [m], [
+        ])
+    ])
+])
+
```
nit: atm AX_PTHREAD produces some output indicating whether pthread was found, but if Win32 is used or there’s no `stdatomic.h` no indicators wrt threading are printed at all. Might be nice to add `AC_MSG_CHECKING` to the Win32 tests or alternatively print the overall result at the end:
```autoconf
thread_api="none" # instead of threads_found
# ....
AC_MSG_NOTICE([Chosen threading model: $thread_api])
```

rcombs replied:
Sure, done.

TheOneric commented on libass/ass_library.c:
```diff
@@ -45,8 +45,13 @@ static void ass_msg_handler(int level, const char *fmt, va_list va, void *data)
         lib->msg_callback = ass_msg_handler;
+#if ENABLE_THREADS
+        lib->thread_safe_cb = -1;
+        pthread_mutex_init(&lib->log_mutex, NULL);
```
What if mutex init fails?

rcombs replied:
Added check.

TheOneric commented on libass/ass_cache.c:
```diff
@@ -17,6 +17,12 @@
+// We have some asserts on highly-contended atomics in here,
+// so they're disabled if not requested
+#if !defined(DEBUG) && !defined(NDEBUG)
+#define NDEBUG 1
```
This disables cache asserts in almost all builds, including our own CI builds. If this is kept, all or most of our CI builds should be changed to explicitly set `-DDEBUG`

rcombs replied:
I think we should either do that, or explicitly put the asserts on newly-atomic values behind an `#ifdef DEBUG` or the like; the perf implications can be pretty major at times, and disabling them in almost all builds is deliberate.

MrSmile replied:
I think instead of locally redefining `NDEBUG`, it would be better to create a specialized time-critical assert function for use in performance-sensitive contexts. It's not exclusive to the cache system, potential tiled rendering can benefit from it too, for example.

rcombs replied:
Went ahead and did this (piggybacking on `DEBUG_LEVEL`).

TheOneric commented on libass/ass_font.c:
```diff
@@ -427,20 +427,23 @@ static int add_face(ASS_FontSelector *fontsel, ASS_Font *font, uint32_t ch)
                     "Got a font face that already is available! Skipping.");
-            return i;
+            ret = i;
+            goto fail;
```
nit: this label is now used for a non-failing condition here; may be better to rename

rcombs replied:
Sure, done.

TheOneric reviewed (commented):
And previously unreleased pending comments from an unfinished review

rcombs commented on libass/ass_library.c:
```diff
@@ -156,6 +164,10 @@ void ass_set_message_cb(ASS_Library *priv,
+#if ENABLE_THREADS
+        if (priv->thread_safe_cb == -1)
+            priv->thread_safe_cb = 0;
+#endif
```
Honestly, I originally brought up this possibility mostly as a strawman to illustrate an argument against locking around these log calls at all, and I mostly regret it; I don't think this is actually a realistic concern, and we've never remotely documented any of the relevant threading behavior, so I'd argue that in the extremely unlikely case that somebody *is* relying on the existing behavior, anything bad that happens to them is their own fault.

TheOneric replied:
I don’t fully agree on the “is their own fault” part, but either way there is no reason to _not_ at least document this change; all i’m asking for is a notice in the commit message (and later the changelog / release notes)

rcombs replied:
Wrote something up; thoughts?

TheOneric replied:
yep, the note added to the “add support for per-event threaded rendering” commit seems good; thanks

rcombs commented:
After a bit of poking around, I have this working in Meson and MSVC (via the Win32 Interlocked* API). Muon is now failing because of the `threads` dep getting applied to nasm for reasons that are completely beyond me (maybe it's just on an old buggy version of meson?), and we don't appear to have the test suite set up for the meson builds yet (definitely would be good to have for this).

Other changes:
- Added `-Watomic-implicit-seq-cst`, avoiding accidental use of arithmetic operators on atomics (these would otherwise silently degrade to unsafe non-atomic accesses on MSVC, which we certainly don't want, and they're also slower).
- Explicitly specified memory ordering everywhere; this should improve perf a bit on ARM.
- Shored up the end-of-frame handling to make sure we can't have any threads go straight from one frame to the next without passing through a full memory barrier.
- Introduced `AtomicInt`, a type suitable for atomic ops (this simplifies the MSVC shims).
- Adjusted cache code to always use the promotion queue for new items, rather than inserting them into the LRU queue during creation; this simplifies the code a bit, and reduces a bit of atomic contention in favor of more use of thread-sharded data; the MRU queue is now entirely non-atomic.
- Always use `atomic_init` to initialize dynamically-allocated atomics (i.e. all of them); this is never actually required on any of our target platforms, but it's essentially free, and is technically required by the standard.
- Update meson build configuration to better match autotools:
  - `_GNU_SOURCE` (a looser mode) instead of `_POSIX_C_SOURCE` (a stricter one, which disabled some functionality we want, here in the form of thread name setters)
  - fribidi bump
  - prefer C11 instead of C99
    - I tried to make this optional, but the Muon build is on too old a Meson to support that; personally I'm fine mandating this, particularly since I'm hoping to make atomics mandatory soon anyway.

arch1t3cht commented on meson.build:
```diff
@@ -135,6 +141,37 @@ if fontconfig
+threads_opt = get_option('threads')
+if not threads_opt.disabled()
+    thread_dep = dependency('threads')
+    if thread_dep.found()
```
Note that `dependency('threads').found()` is currently [always true](https://github.com/mesonbuild/meson/blob/7ec1fc507ea6f80e819ac46b56125b955073de7f/mesonbuild/dependencies/misc.py#L149), even if the compiler does not actually support the `-pthread` argument (see also https://github.com/mesonbuild/meson/issues/7775).
Of course, it's always possible for this to change in the future, but I wanted to point this out in case you weren't aware.

rcombs replied:
Good to know, though I'm not sure if there's much to be done about it at the moment.

TheOneric commented:
> muon is now failing because of the threads dep getting applied to nasm for reasons that are completely beyond me (maybe it's just on an old buggy version of meson?),

We do (now) use a somewhat old 0.2.0 in CI, but this also happens in current git master. I filed an issue upstream: https://todo.sr.ht/~lattis/muon/129

rcombs commented:
@debugzxcv I don't believe that crash can possibly occur on the current version of the branch, but it's hard to tell without source line information, or at least a disassembly of the crashing function. What git hash are you running?

TheOneric commented:
> [7d5208a](https://github.com/libass/libass/commit/7d5208a6eccf3b83dabfc2cf497e9a4dd5b25268) + this branch

What commit from this PR are you using? This branch keeps changing

kasper93 commented:
> which makes mpv stats.lua actually usable without stuttering.

It doesn't solve anything.

TheOneric commented on meson.build:
```diff
@@ -71,6 +72,11 @@ if (
     conf.set('HAVE_FSTAT', 1)
 endif
 
+have_stdatomic = cc.check_header('stdatomic.h')
+if have_stdatomic
+    conf.set('HAVE_STDATOMIC_H', 1)
+endif
```
`have_stdatomic` is used later on and thus cannot be inlined into `conf.set()`, but it’ſ true the other two checks/defines are currently missing in meson

MrSmile commented on libass/ass_render.c:
```diff
@@ -3049,6 +3130,67 @@ ass_render_event(RenderContext *state, ASS_Event *event,
+    pthread_cond_broadcast(&priv->main_cond);
+
+    for (;;) {
+        while (!priv->shutting_down && atomic_load_explicit(&priv->next_eimg, memory_order_relaxed) >= atomic_load_explicit(&priv->sent_eimgs, memory_order_relaxed))
```
Too long line.

MrSmile commented on libass/ass_render.c:
```diff
@@ -3374,6 +3505,42 @@ static int ass_detect_change(ASS_Renderer *priv)
+#if ENABLE_THREADS
+static bool start_threads(ASS_Renderer *priv)
+{
+    if (priv->n_threads != priv->settings.threads && !priv->thread_start_failed) {
```
Excess indentation. I think it's better to use
```c
     if (priv->n_threads == priv->settings.threads || priv->thread_start_failed)
        return true;
```

MrSmile commented on libass/ass_render.h:
```diff
@@ -342,6 +349,21 @@ struct ass_renderer {
+
+#if ENABLE_THREADS
+    pthread_mutex_t mutex;
+    int mutex_inited;
+    pthread_cond_t main_cond, pool_cond;
+    int main_cond_inited, pool_cond_inited;
+
+    unsigned n_threads, started_threads, active_threads;
+    pthread_t *threads;
+
+    int thread_start_failed;
```
Failure modes are total mess here: 7 status variables instead of 1&ndash;2. I feel it can be greatly simplified.

MrSmile reviewed (commented):
I think that ideal solution for logging would be to pin in to the main thread only. But that would be quite substantial and intrusive change without much return...

rcombs commented on libass/ass_cache.c:
```diff
@@ -461,80 +661,115 @@ void ass_cache_inc_ref(void *value)
+        CacheItem *item_next = atomic_load_explicit(&item->next, memory_order_relaxed);
+        if (item_next)
+            item_next->prev = item->prev;
+        atomic_store_explicit(item->prev, item_next, memory_order_relaxed);
```
That isn't really actionable on its own; do you have a full all-thread stack trace? The rest of the state of `*item`?

rcombs replied:
I'm not sure if it's possible to debug with a pdb but no corresponding executable? (If there is, I don't know how to do it)

rcombs replied:
Can you reproduce this crash routinely, or is it rare and random? Does it happen with a particular sample file?

I've added some new debug assertions that may help catch this kind of issue sooner; if you pull the latest changes and build with `-DDEBUG_LEVEL=2`, hopefully you'll hit an assert before a crash like this happens, and the dump from that earlier crash should be useful in diagnosing the problem.

Side-note, what build system and compiler did you use to build the libass used in that dump? I noticed that the atomic variables behave somewhat strangely in the debug symbols; this probably isn't related to the cause of the crash, but it'd be useful to know what causes that.

rcombs replied:
I've pushed a change that I believe should fix this. It was reproducible by rendering multiple frames that each produced only one bitmap (i.e. a single event with a single line with `{\bord0\shad0}`). Thanks for the report; please confirm if it still recurs on the latest version of the branch.

rcombs committed:
ass_utils: add ass_assert[1,2] macros

rcombs committed:
ass_render: support overriding default cache sizes via env var

rcombs commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,348 @@
+#include <stdatomic.h>
+#define HAVE_ATOMICS 1
+
+typedef intptr_t AtomicInt;
```
I'd kinda prefer that all else being equal, but the Windows atomic functions used by the compat shims all operate on `LONG`/`LONG64`, which are signed, so I picked `intptr_t` for consistency with those.

rcombs commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,348 @@
+static inline bool INTERLOCKED(DoCompareExchangePointer, order)(void *obj, void *expected, void *desired) \
+{ \
+    void *old = *(void**)expected; \
+    return (*(void**)expected = INTERLOCKED(CompareExchangePointer, order)((void**)obj, expected, desired)) == old; \
```
We have a CI job that builds for win64 and it does work, but you have a point; I have no idea *why* this works. Changed to just do `InterlockedCompareExchangePointer ## order`.

astiob replied:
Could it be that our CI has stdatomic.h?

astiob replied:
It seems [there was](https://github.com/libass/libass/actions/runs/12193986212/job/34017107542#step:10:316) an undefined-symbol warning:

```
'InterlockedCompareExchangePointer64' undefined; assuming extern returning int
```

It looks like in the end it only builds a static `libass.a`, so I guess the presence of external symbols isn’t verified. @TheOneric?

(Meanwhile in MinGW-w64 builds, stdatomic.h is indeed present.)

rcombs replied:
Oh, do we not have tests in the meson builds yet?

TheOneric replied:
> do we not have tests in the meson builds yet?
nope, meson doesn’t even know how to build `compare`, `fuzz` or `checkasm`

Though we probably should switch the meson windows builds to `default_library=both` to make sure linking works. This should also catch undefined symbols though it’d be good to enable error and warning options in MSVC similar to the set we enable for GCC-like compilers if anyone knows which options MSVC has for this.

rcombs commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,348 @@
+
+#ifdef ENABLE_THREADS
+
+#define inc_ref(x) (void) (atomic_fetch_add_explicit(x, 1, memory_order_relaxed))
+#define dec_ref(x) (atomic_fetch_sub_explicit(x, 1, memory_order_acq_rel) - 1)
```
These are meant specifically for refcounts, where release is correct for incrementing and acquire-release is correct for decrementing. Other usage can call the underlying functions directly.

rcombs commented on libass/ass_shaper.c:
```diff
@@ -425,25 +452,42 @@ get_contour_point(hb_font_t *font, void *font_data, hb_codepoint_t glyph,
 {
-    FT_Face face = font->faces[index];
-    hb_face_t *hb_face = hb_face_create_for_tables(get_reference_table, face, NULL);
+    struct font_ref *key = malloc(sizeof(struct font_ref));
```
Seems like a premature optimization to me; this only runs once per font face, it's not hot-path by any stretch of the imagination.

rcombs commented on libass/ass_cache.c:
```diff
@@ -377,11 +381,55 @@ Cache *ass_cache_create(const CacheDesc *desc)
+{
+    CacheClient *client = NULL;
+
+    client = calloc(1, sizeof(*client));
```
Just good practice in general; the extra cost of a cold-path `calloc` vs `malloc` is extremely negligible.

rcombs committed:
ass_cache: rename Constructor->Construct for consistency

rcombs committed:
profile: log timing data

rcombs committed:
profile: support setting log level

rcombs commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,348 @@
+#define IS_RELAXED(order) ((#order)[0] == 'N')
+#define MAYBE_FENCE(order) (!IS_RELAXED(order) ? MemoryBarrier() : (void)0)
+
+#define DoCompareExchangeStrong(order) \
```
Capitalized; not `#undef`-ing since these are internal and the names are pretty specific to atomics.

rcombs commented on libass/ass_cache.h:
```diff
@@ -52,7 +52,8 @@ typedef ass_hashcode (*HashFunction)(void *key, ass_hashcode hval);
 typedef size_t (*CacheValueConstructor)(void *key, void *value, void *priv);
-typedef void (*CacheItemDestructor)(void *key, void *value);
+typedef void (*CacheValueDestructor)(void *value);
+typedef void (*CacheKeyDestructor)(void *key);
```
Done.

TheOneric commented on meson.build:
```diff
@@ -52,7 +53,7 @@ add_project_arguments(
 # Configuration
 
-str_check_functions = ['strdup', 'strndup']
+str_check_functions = ['strdup', 'strndup', 'sched_getaffinity']
```
this won't work; functions in this array are cheked for existence _in the `string.h` header_

rcombs commented:
This is now rebased on latest upstream master; it's building on MSVC, and the tests are passing (getting the utils built did turn up a couple issues, which were trivial to fix; the tests themselves passed without any further changes).

rcombs committed:
profile: default to a less-noisy log level

rcombs commented:
Not sure why the MSVC build is broken after a rebase; my guess is it's something to do with the changes to the build configuration that have recently merged. Patches welcome from windows-touchers, or I can just turn threading back off for MSVC builds.

rcombs commented:
This was an entirely legitimate use-before-init on a mutex; no idea why it didn't crash previously, or was never caught by any of our sanitizers. Fixed; thanks much for the stack.

kasper93 commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,360 @@
+static inline int pthread_mutex_init(pthread_mutex_t *mtx, const void *attr)
+{
+    assert(!attr);
+    InitializeCriticalSection(mtx);
```
If recursive mutexes are not needed, consider using SWR locks instead. They are lighter and faster.

rcombs replied:
Interesting; it looks like SRW locks perform better in *most* (but not all) cases in this benchmark, but the results appear fairly noisy, and are highly dependent on exactly how much contention is happening. I'd still definitely be curious to see a measured impact in a real-world libass scenario.

ThaUnknown commented:
<img width="1468" height="501" alt="image" src="https://github.com/user-attachments/assets/a54dcdb1-67e1-4382-986b-cb687250fb0a" />

works as advertised

Andarwinux commented:
Yes, this allows fancy special effects subtitles to be rendered at an acceptable speed without any dropped frames on CPUs without excellent single-threaded performance. Unfortunately, this PR seems to have stalled.

ThaUnknown commented:
I think I found a problem

Please take this bug report with a heavy grain of salt, I might be doing something majorly wrong, as my implementation is quite "non-standard". So most of the things here are a "I think".

There appears to be fringing happening on the first character, symbol or glyph of each line. This happens on the start of the character/symbol usually only for the first 5-ish pixels, then everything else continues normally. If there are multiple lines in the presented subtitle, and they don't line up this issue shows up on both lines, so its like not a "5px of the left side of the bitmap" but "5px of the left side of the text line", I've put together a collage which shows this happening in many characters, symbols etc, all of which I doubled to show where the problem is visible:
<img width="374" height="302" alt="image" src="https://github.com/user-attachments/assets/a3a26f49-8f5e-4cec-8649-2eee7af4c3cd" />
It would be nice if some1 else could reproduce the issue, from what I can tell the font that causes this doesn't matter, but I'm not sure if things like outline etc matter, I can provide extra information if this is non-reproducible, but for me it happens 100% of the time and is 100% reproducible and consistent.


There's a non-0 chance this is also a bug in my implementation, as I'm using experimental vulkan features, an intel arc gpu and some quite stupid hacks to render this, but from my albeit limited testing I don't think that's the case.


Another problem I found is the thread count allocation, this could entirely be an emscripten/wasm bug, but I was not able to find information on this being the case so I'll report it just in case:
The thread count allocation seems weird, if you call `ass_set_threads` with `1`, it will only use the one main thread and no extra threads, but calling it with `2` makes it want to use the main thread +2 pthreads, which means `3`, which means you can never use 2 threads for rendering, only 1, 3, 4, 5 etc, also the entire library froze up when it wanted to spawn a pthread but there were no more available threads, but this could be a wasm problem

rcombs commented:
Re: the antialiasing oddities in that screenshot, I can't repro locally, and all test cases pass here (as well as in the CI). Does the test suite pass in your build configuration? To ask the obvious question, can you reproduce this *without* the "experimental vulkan features" and "stupid hacks"?

Re: thread count, you're describing the intended behavior. Setting the thread count to 2 means that 2 threads will be used for rendering (while the thread in `ass_render_frame` simply waits for them to finish). This is standard pool-size-configuration behavior; the point isn't to request that a particular number of total threads *exist*, but instead to request that no more than a particular number of cores are loaded at any given time.

> the entire library froze up when it wanted to spawn a pthread but there were no more available threads

What does it mean for there to be "no more available threads"? What is `pthread_create` returning? When you say the library "froze up", which call is blocking?

ThaUnknown commented:
Welp then just ignore me then if you can't reproduce this issue, I was hopeful that it wasn't something on my end but it must be.

> Re: the antialiasing oddities in that screenshot, I can't repro locally, and all test cases pass here (as well as in the CI). Does the test suite pass in your build configuration?

Ah you see here in 'ye ole' WASM land things aren't so simple, so I don't really know.don

> To ask the obvious question, can you reproduce this _without_ the "experimental vulkan features" and "stupid hacks"?

If it was easy I'd have done it in the first place, I spent 4 days making the last renderer, 2 of which I didn't sleep so I'll assume it's some bug in my pipeline somewhere.

> Re: thread count, you're describing the intended behavior. Setting the thread count to 2 means that 2 threads will be used for rendering (while the thread in `ass_render_frame` simply waits for them to finish). This is standard pool-size-configuration behavior; the point isn't to request that a particular number of total threads _exist_, but instead to request that no more than a particular number of cores are loaded at any given time.

Ah that make sense, unfortunate since it wastes quite a bit of resources in WASM land, but yeah if that's the norm I understand

> What does it mean for there to be "no more available threads"? What is `pthread_create` returning? When you say the library "froze up", which call is blocking?

When running WASM with pthreads for emscripten you specify how many pthreads it can create, so exceeding that value hangs. I can give you the WASM call id if you want to, but that's useless, can't do much more than that since libass+DWARF doesn't rly go well together idk why.

ThaUnknown commented:
yeah, sorry for spamming this thread with useless comments, the code from this PR works flawless, I fucked up floating point precision in the vulkan shader, my bad!

MrSmile commented on profile/profile.c:
```diff
@@ -62,6 +68,20 @@ static void init(int frame_w, int frame_h)
+#else
+    struct timespec spec;
+    clock_gettime(CLOCK_MONOTONIC, &spec);
+    return spec.tv_sec + spec.tv_nsec / 1000000000.;
```
```c
return spec.tv_sec + spec.tv_nsec * 1e-9;
```
Division is slow and compilers cannot replace it with multiplication without `-ffast-math` (but they happily replace `a + b * c` with FMA with about the same discrepancy, where's logic in it?).

Not sure about returning floating point from timing functions though, I generally consider that a bad practice. Maybe it's better to use a more precise approach with separate multiplier applied later:
```c
static double get_time_factor(void)
{
#ifdef _WIN32
    LARGE_INTEGER frequency;
    QueryPerformanceFrequency(&frequency);
    return 1.0 / frequency.QuadPart;
#else
    return 1e-9;
#endif
}

static uint64_t get_time(void)
{
#ifdef _WIN32
    LARGE_INTEGER counter;
    QueryPerformanceCounter(&counter);
    return counter.QuadPart;
#else
    struct timespec spec;
    clock_gettime(CLOCK_MONOTONIC, &spec);
    return 1000000000 * spec.tv_sec + spec.tv_nsec;
#endif
}
```

TheOneric replied:
I like the `get_factor_version`. To avoid potential precision loss though, we could return the integer value directly as `uint64_t get_time_denominator(void)`

rcombs replied:
Is this realistically ever going to matter at the precisions we're working with here?

MrSmile commented on profile/profile.c:
```diff
@@ -102,6 +104,11 @@ int main(int argc, char *argv[])
         exit(1);
     }
 
+    const char *log_level_var = getenv("LIBASS_LOG_LEVEL");
```
Maybe command line parameter is better?

MrSmile commented on libass/ass_utils.h:
```diff
@@ -55,6 +55,18 @@
+#if DEBUG_LEVEL >= 1
+    #define ass_assert1(x) assert(x)
+#else
+    #define ass_assert1(x) (void)0
```
Maybe it's safer to add extra brackets `((void) 0)`?
Also I think it's better to be consistent with libass macro style and name it `ASSERT1`.

MrSmile commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,360 @@
+
+#define ENABLE_THREADS 0
+
+#define pthread_mutex_lock(x) do {} while(0)
+#define pthread_mutex_unlock(x) do {} while(0)
+#define pthread_cond_signal(x) do {} while(0)
+#define pthread_cond_broadcast(x) do {} while(0)
```
Earlier for no-op operations were used `((void) 0)`, better to be consistent.

MrSmile commented on libass/ass_font.c:
```diff
@@ -498,17 +498,27 @@ size_t ass_font_construct(void *key, void *value, void *priv)
-    if (error == -1)
+    if (error == -1) {
         font->library = NULL;
+#if ENABLE_THREADS
+        pthread_mutex_destroy(&font->mutex);
+#endif
```
Maybe it's better to create no-op stub for `pthread_mutex_destroy` too?
Possibly even for `pthread_mutex_init`.

MrSmile commented on libass/ass_cache.c:
```diff
@@ -453,59 +510,117 @@ void *ass_cache_get(Cache *cache, CacheClient *client, void *key, void *priv)
+
+    if (item != NULL && item != stop_at) {
+        if (atomic_load_explicit(&item->last_used_frame, memory_order_consume) != cache->cur_frame) {
+            uintptr_t last_used = atomic_exchange_explicit(&item->last_used_frame, cache->cur_frame, memory_order_consume);
```
`memory_order_consume` is considered somewhat problematic, so maybe it's better to use `memory_order_acquire` instead?

rcombs replied:
Hmmm, it looks like `memory_order_consume` was deprecated in C++, and has never been implemented anywhere (it's effectively identical to `memory_order_acquire` in all real implementations), so sure, might as well make that explicit here.

MrSmile commented on libass/ass_cache.c:
```diff
@@ -453,59 +510,117 @@ void *ass_cache_get(Cache *cache, CacheClient *client, void *key, void *priv)
-
-    cache->cache_size += item->size + (item->size == 1 ? 0 : CACHE_ITEM_SIZE);
+    size_t size = desc->construct_func(new_key, value, priv);
+    ass_assert1(size);
```
This assert is slow? Looks like it should be dominated by indirect function call earlier.

rcombs replied:
I don't expect it to be particularly slow; I'm just trying to avoid 0-level asserts in cases where failure is still *safe*.

MrSmile commented on libass/ass_cache.c:
```diff
@@ -453,59 +510,117 @@ void *ass_cache_get(Cache *cache, CacheClient *client, void *key, void *priv)
+
+    atomic_fetch_add_explicit(&cache->cache_size, size + (size == 1 ? 0 : CACHE_ITEM_SIZE), memory_order_relaxed);
+
+#if ENABLE_THREADS
```
`#if` should be unnecessary.

TheOneric replied:
`client->mutex` only exists if `ENABLE_THREADS` is defined thruthy.
*EDIT:* ah, reading your other comments, you probably mean it won’t be necessary anymore with a noop mutex creation/destruction macro for threadless builds

rcombs replied:
Personally, I think it's a little icky to have code that appears to reference struct members that might not exist (and then just have it not be compiled because it's a macro arg).

MrSmile commented on libass/ass_cache.c:
```diff
@@ -453,59 +510,117 @@ void *ass_cache_get(Cache *cache, CacheClient *client, void *key, void *priv)
+
+    atomic_store_explicit(&item->size, size, memory_order_release);
+
+#if ENABLE_THREADS
```
`#if` should be unnecessary.

MrSmile commented on libass/ass_cache.c:
```diff
@@ -531,81 +647,134 @@ void ass_cache_inc_ref(void *value)
     CacheItem *item = value_to_item(value);
-    assert(item->size && item->ref_count);
-    item->ref_count++;
+    ass_assert2(atomic_load_explicit(&item->size, memory_order_acquire) && atomic_load_explicit(&item->ref_count, memory_order_acquire));
```
Better to split into two individual asserts.

MrSmile commented on libass/ass_cache.c:
```diff
@@ -531,81 +647,134 @@ void ass_cache_inc_ref(void *value)
+
+static void dec_ref_item(CacheItem *item)
+{
+    ass_assert2(atomic_load_explicit(&item->size, memory_order_acquire) && atomic_load_explicit(&item->ref_count, memory_order_acquire));
```
Here too.

MrSmile commented on libass/ass_threading.h:
```diff
@@ -0,0 +1,360 @@
+
+#endif /* CONFIG_PTHREAD */
+
+#define inc_ref(x) (void) (atomic_fetch_add_explicit(x, 1, memory_order_relaxed))
+#define dec_ref(x) (atomic_fetch_sub_explicit(x, 1, memory_order_acq_rel) - 1)
```
Not sure about usefulness of this. I think it's better to keep memory order explicit.

rcombs replied:
The memory ordering required for refcounting is always the same; having it specified in a single consistent place avoids confusion.

MrSmile reviewed (commented):
Mostly cosmetic nits. Substantial code should be fairly tested by now, I think. What do we do about pathological cases (memory bound) though? Ignore for now?

TheOneric commented on profile/profile.c:
```diff
@@ -82,21 +102,66 @@ int main(int argc, char *argv[])
+    printf("   total render: %f\n", last_frame_time - read_time);
+    if (frames > 0) {
+        printf("    first frame: %f\n", first_frame_time - read_time);
+        printf("     post-first: %f\n", last_frame_time - first_frame_time);
```
as is, this seems redundant with the already logged “total render” and “first frame”. Instead I think it would be good to log the count of rendered frames (perhaps differentiating empty frames without any `ASS_Image`s from those with actual results)

TheOneric replied:
this is what i’m thinking of *(`average frame` is just include for a convenient comparison; strictly speaking it’s redundant)*:
```diff
diff --git a/profile/profile.c b/profile/profile.c
index e8aed7f2..89955d8a 100644
--- a/profile/profile.c
+++ b/profile/profile.c
@@ -128,6 +128,7 @@ int main(int argc, char *argv[])
     double worst_frame = 0;
     bool got_frame = false;
     int frames = 0;
+    int nonempty_frames = 0;
 
     while (tm < end_time) {
         bool got = ass_render_frame(ass_renderer, track, (int) (tm * 1000), NULL) != NULL;
@@ -146,6 +147,7 @@ int main(int argc, char *argv[])
         last_frame_time = current_time;
 
         frames++;
+        if (got) nonempty_frames++;
     }
 
     ass_free_track(track);
@@ -154,14 +156,17 @@ int main(int argc, char *argv[])
 
     double cleanup_time = gettime();
 
-    printf("Timing:\n");
+    printf("Timing [seconds]:\n");
     printf("           init: %f\n", init_time - start_time);
     printf("           read: %f\n", read_time - init_time);
     printf("   total render: %f\n", last_frame_time - read_time);
     if (frames > 0) {
         printf("    first frame: %f\n", first_frame_time - read_time);
-        printf("     post-first: %f\n", last_frame_time - first_frame_time);
         printf("    worst frame: %f\n", worst_frame);
+        printf("  average frame: %f\n", (last_frame_time - read_time) / frames);
+        printf("     all frames: %f\n", last_frame_time - read_time);
+        printf("total frame cnt: %d\n", frames);
+        printf("nonempty frames: %d\n", nonempty_frames);
     }
     printf("        cleanup: %f\n", cleanup_time - last_frame_time);
     if (frames > 0) {

```

TheOneric commented on profile/profile.c:
```diff
@@ -82,21 +102,66 @@ int main(int argc, char *argv[])
+    if (frames > 0) {
+        printf("      total fps: %f\n", frames / (last_frame_time - read_time));
+        printf("     post-1 fps: %f\n", frames / (last_frame_time - first_frame_time));
+        printf("     post-1 fps: %f\n", frames / (last_frame_time - first_frame_time));
```
“post-1 fps” log is duplicated

TheOneric commented on profile/profile.c:
```diff
@@ -35,9 +41,9 @@ void msg_callback(int level, const char *fmt, va_list va, void *data)
-    printf("\n");
+    char fmt_buf[1024];
+    snprintf(fmt_buf, sizeof(fmt_buf), "libass: %s\n", fmt);
+    vfprintf(stderr, fmt_buf, va);
```
The `snprintf` result is used unchecked (and thus might miss the terminating zero byte).
Given it only needs to accommodate _format_ strings from our own code and `assert` might suffice, but there should be some form of check

rcombs replied:
> and thus might miss the terminating zero byte

Not sure what this means? `snprintf` always 0-terminates its output even in the insufficient-space case. Might miss the trailing newline, though.

TheOneric replied:
you’re right, my bad. Mixed it up with other APIs like `strncpy`.
Might still be nice to have something loudly complain in case the buffer ever ends up too small, so we can bump it. But nothing critical

TheOneric reviewed (commented):
The first two commits could already have been merged much earlier like the other prep work, wonder why we didn’t do that yet

ThaUnknown commented:
would it be somehow possible to reduce the amt of calls for system time when WebAssembly build target is detected? they are quite expensive since they cross the WASM boundary, that said I don't know if that even is possible with the current architecture of this PR

other than that I've been running this PR in my software, and after a few hundred thousand hours of usage no issues arose, good job on this

TheOneric commented:
if you’re talking about calls to the `get_time` function discussed in preceding set of comments, they are part of the `profile` utility not the library itself. Thus I don’t think performance of it in WASM amtters; unless you want to use the profile utility inside WASM, in which case you’re welcome to contribute a more efficient way to get execution times in WASM builds. But even then this doesn’t seem like a blocker to me

Andarwinux commented:
To improve WASM performance, it may be necessary to add WASM simd128 asm. On x86, even with clang avx512 automatic vectorization, the C implementation in libass is still extremely slow.

ThaUnknown commented:
> calls to the get_time function

no, I _**think**_ SleepConditionVariableSRW calls system time, do it enough times and it stacks up

>  you’re welcome to contribute a more efficient way to get execution times in WASM builds

I wish I was good enough at C to do that, I'd have submitted a PR to this PR by now, unfortunately I struggled, so I came here to ask if reducing the amt of system calls for time even is possible at all

> improve WASM performance

[libass in WASM performs fine](https://jassub.pages.dev/beastars/jassub/), this was strictly about the [relatively] frequent system time calls when a LOOOT of rendering is being done

rcombs commented:
Well given that SleepConditionVariableSRW is a WinAPI function, it seems *pretty* unlikely that it would be called in a wasm build. Can we take the unrelated chatter elsewhere, please?

TheOneric commented on profile/profile.c:
```diff
@@ -82,21 +105,70 @@ int main(int argc, char *argv[])
+    printf("        cleanup: %f\n", cleanup_time - last_frame_time);
+    if (frames > 0) {
+        printf("      total fps: %f\n", frames / (last_frame_time - read_time));
+        printf("     post-1 fps: %f\n", frames / (last_frame_time - first_frame_time));
```
It’s a false warning since it’s gated behind `frames  > 0`, but my compiler complains about `first_frame_time` being used here while being possibly uninitialised.  
Though given `last_frame_time` also already gets a default initialisation, we could just initialise `first_frame_time` too for consistency avoiding the warning
